ballerina/modules/parser/lexer.bal

```diff
@@ -103,22 +103,124 @@ public class Lexer {
+                        word += value;
+                    }
+                    previousChar = value;
+                }
+            }
+            return getToken(word, T_STRING, location);
+        }
+    }
+
+    isolated function readBlockStringLiteral() returns string {
+        string previousChar = "";
+        string word = "";
+        string[] lines = [];
+        _ = self.readNextChar(); // Ignore third double quote character
+        while true {
             string value = self.charReader.peek();
-            if (value is LineTerminator) {
-                string message = "Syntax Error: Unterminated string.";
-                return error UnterminatedStringError(message, line = self.currentLocation.line,
-                                                     column = self.currentLocation.column);
-            } else if (value == DOUBLE_QUOTE && previousChar != BACK_SLASH) {
+            if value == NEW_LINE {
+                lines.push(word);
+                word = "";
                 value = self.readNextChar();
-                break;
+            } else if value == DOUBLE_QUOTE {
+                string wordWithQuotes = self.readQuotesInBlockStrings(word);
+                if wordWithQuotes == word {
+                    lines.push(word);
+                    break;
+                } else {
+                    word = wordWithQuotes;
+                }
             } else {
                 value = self.readNextChar();
                 word += value;
             }
             previousChar = value;
         }
-        return getToken(word, T_STRING, location);
+        return self.getBlockStringValue(lines);
+    }
+
+    isolated function readQuotesInBlockStrings(string word) returns string {
+        string previousChar = self.readNextChar();
+        string value = self.charReader.peek();
+        string returnString = word;
+        if value == DOUBLE_QUOTE {
+            previousChar = self.readNextChar();
+            value = self.charReader.peek();
+            if value == DOUBLE_QUOTE {
+                value = self.readNextChar();
+                if word.endsWith(BACK_SLASH) {
+                    returnString = word.substring(0, word.length()-1);
+                    returnString += value;
+                } else {
+                    return returnString;
+                }
+            }
+            returnString += previousChar;
+        }
+        returnString += previousChar;
+        return returnString;
+    }
+
+    isolated function getBlockStringValue(string[] lines) returns string {
+        int? commonIndent = ();
+        string formatted = "";
+        foreach string line in lines {
+            int indent  = self.getLeadingWhiteSpaceCount(line);
+            if indent < line.length() && (commonIndent is () || indent < commonIndent) {
+                commonIndent = indent;
+            }
+        }
+        if commonIndent is int {
+            foreach string line in lines {
```

Comment:
Since we need the index of the line inside the loop, better to use a for loop with a number range instead of a foreach loop. 
[Check this BBE](https://ballerina.io/learn/by-example/foreach-statement.html?is_ref_by_example=true)